File: app/src/main/java/com/babestudios/companieshouse/data/network/converters/ToStringConverterFactory.java
package com.babestudios.companieshouse.data.network.converters;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

public class ToStringConverterFactory extends Converter.Factory {
	private static final MediaType MEDIA_TYPE = MediaType.parse("text/plain");

	@Override
	public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations, Retrofit retrofit) {
		if (String.class.equals(type)) {
			return value -> value.string();
		}
		return null;
	}

	@Override
	public Converter<?, RequestBody> requestBodyConverter(Type type, Annotation[] parameterAnnotations, Annotation[] methodAnnotations, Retrofit retrofit) {
		if (String.class.equals(type)) {
			return new Converter<String, RequestBody>() {
				@Override
				public RequestBody convert(String value) throws IOException {
					return RequestBody.create(MEDIA_TYPE, value);
				}
			};
		}
		return null;
	}

	public static ToStringConverterFactory create() {
		return new ToStringConverterFactory();
	}
}

File: data/src/main/java/com/babestudios/companyinfouk/data/network/converters/ToStringConverterFactory.kt
package com.babestudios.companyinfouk.data.network.converters

import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Converter
import retrofit2.Retrofit
import java.lang.reflect.Type

class ToStringConverterFactory : Converter.Factory() {

	override fun responseBodyConverter(
			type: Type?,
			annotations: Array<Annotation>?,
			retrofit: Retrofit?
	): Converter<ResponseBody, String>? {
		if (String::class.java == type) {
			return Converter { value -> value.string() }
		}
		return null
	}

	override fun requestBodyConverter(
			type: Type?,
			parameterAnnotations: Array<Annotation>?,
			methodAnnotations: Array<Annotation>?,
			retrofit: Retrofit?
	): Converter<*, RequestBody>? {
		return if (String::class.java == type) {
			Converter<String, RequestBody> { value -> RequestBody.create(MEDIA_TYPE, value) }
		} else null
	}

	companion object {
		private val MEDIA_TYPE = "text/plain".toMediaTypeOrNull()

		fun create(): ToStringConverterFactory {
			return ToStringConverterFactory()
		}
	}
}
